libdnf/utils/utils.cpp

```diff
@@ -311,6 +312,74 @@ void decompress(const char * inPath, const char * outPath, mode_t outMode, const
     fclose(inFile);
 }
 
+bool checksum_check(const char * type, const char * inPath, const char * checksum_valid)
+{
+    GError * errP{nullptr};
+    gboolean valid;
+    LrChecksumType lr_type = lr_checksum_type(type);
+
+    if (lr_type == LR_CHECKSUM_UNKNOWN)
+        throw libdnf::Error(tfm::format("Unknown checksum type %s", type));
+
+    auto inFd = open(inPath, O_RDONLY);
+
+    if (inFd == -1)
+        throw libdnf::Error(tfm::format("Error opening %s: %s", inPath, strerror(errno)));
+
+    auto ret = lr_checksum_fd_cmp(lr_type,
+                      inFd,
+                      checksum_valid,
+                      TRUE, /* use xattr value */
+                      &valid,
+                      &errP);
+
+    if (!ret) {
+      close(inFd);
+      throw libdnf::Error(tfm::format("Error checking checksum %s: (%d, %s)", inPath, errP->code, errP->message));
+    }
+    close(inFd);
+    // gboolean -> bool
+    return valid == TRUE;
+}
+
+std::string checksum_value(const char * type, const char * inPath)
+{
+    GError * errP{nullptr};
+    gboolean valid; /* this is thrown away, we're not checking, we're just calculating it */
+    gchar * calculated = NULL;
+    LrChecksumType lr_type = lr_checksum_type(type);
+
+    if (lr_type == LR_CHECKSUM_UNKNOWN)
+        throw libdnf::Error(tfm::format("Unknown checksum type %s", type));
+
+    auto inFd = open(inPath, O_RDONLY);
+
+    if (inFd == -1)
+        throw libdnf::Error(tfm::format("Error opening %s: %s", inPath, strerror(errno)));
+
+    auto ret = lr_checksum_fd_compare(lr_type,
+                      inFd,
+                      "",   /* we're calling this to get the file's checksum not compare it */
+                      /**
+                       * https://github.com/rpm-software-management/librepo/issues/233
+                       * We must pass FALSE here because librepo has a simple bug that
+                       * doesn't set calculated if the checksum is cached
+                       */
+                      FALSE,
+                      &valid,
+                      &calculated,
+                      &errP);
+
+    if (!ret) {
+      close(inFd);
+      throw libdnf::Error(tfm::format("Error calculating checksum %s: (%d, %s)", inPath, errP->code, errP->message));
+    }
+    close(inFd);
+    std::string out(calculated);
+    g_free(calculated);
+    return out;
+}
+
```

Comment:
There is a lot of duplicated code - both functions are very similar. It should be possible to deduplicate the code to auxiliary function which would be than used by both functions.
Even `lr_checksum_fd_cmp` could be replaced with `lr_checksum_fd_compare` so the only remaining differences in both functions is the return value and some of `lr_checksum_fd_compare` parameters values.

Reply:
I had a go at this initially and failed to come up with something nice, but now I think i got it.